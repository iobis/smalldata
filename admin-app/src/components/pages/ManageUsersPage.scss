.manage-users-page {
  .users {
    margin-top: var(--margin-top);

    .user-row {
      .datasets {
        ul {
          list-style-type: disc;
        }
      }
    }
  }
}

a.button.new {
  background-color: var(--obis-green);
  margin-bottom: var(--margin-bottom);

  &:hover {
    background-color: lighten($obis-green, 10%);
  }
}
