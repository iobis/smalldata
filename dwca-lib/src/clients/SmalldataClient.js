import { mapDatasetToRequest, mapOccurrenceToDwca } from './SmalldataConverters'

const authorizationValue = 'Basic verysecret'
const headers = {
  'Authorization': authorizationValue,
  'Content-Type':  'application/json'
}

export async function getDatasets() {
  const response = await fetch('/api/datasets', { headers })
    .then(response => response.json())
  return response.map(renameRefToId)
}

export async function createDataset(dataset) {
  const request = mapDatasetToRequest(dataset)
  return await fetch('/api/datasets', {
    method: 'POST',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
}

export function renameRefToId({ ref, ...rest }) {
  return ({ id: ref, ...rest })
}

export async function getOccurrences({ userRef }) {
  const query = 'projectFields=dwcRecord.tdwg.scientificName&projectFields=dwcRecord.tdwg.eventDate'
  return await fetch(`/api/dwca/user/${userRef}/records?${query}`, { headers })
    .then(response => response.json())
}

export async function getOccurrence({ datasetId, dwcaId, userRef }) {
  return fetch(`/api/dwca/${datasetId}/user/${userRef}/records/${encodeURIComponent(dwcaId)}`, { headers })
    .then(response => response.json())
}

export async function findLatestOccurrence({ userRef }) {
  const occurrences = await getOccurrences({ userRef })
  const latest = occurrences
    .filter(dwca => !!dwca.addedAtInstant)
    .sort((left, right) => left.addedAtInstant < right.addedAtInstant ? 1 : -1)[0]
  return await getOccurrence({ userRef, dwcaId: latest.dwcaId, datasetId: latest.dataset })
}

export async function updateOccurrence({ occurrence, userRef, dwcaId }) {
  const request = mapOccurrenceToDwca(occurrence)
  const url = `/api/dwca/${occurrence.dataset.id}/user/${userRef}/records/${encodeURIComponent(dwcaId)}`
  return await fetch(url, {
    method: 'PUT',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
}

export async function createOccurrence({ occurrence, userRef }) {
  const request = mapOccurrenceToDwca(occurrence)
  const url = `/api/dwca/${occurrence.dataset.id}/user/${userRef}/records`
  return await fetch(url, {
    method: 'POST',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
}

export async function getUsers() {
  return fetch('/api/users', { headers })
    .then(response => response.json())
}

export async function getUsersWithDatasets() {
  const [users, datasets] = await Promise.all([getUsers(), getDatasets()])
  const datasetIdToDataset = groupBy(datasets, 'id')
  return users.map(user => {
    if (!user['dataset_refs']) user['dataset_refs'] = []
    const datasets = user['dataset_refs'].map(datasetRef => datasetIdToDataset[datasetRef][0])
    return { ...user, datasets }
  })
}

export async function createUser({ datasetIds, email, name, role }) {
  const request = {
    'dataset_refs': datasetIds,
    emailAddress:   email,
    name,
    role
  }
  const url = '/api/users'
  return await fetch(url, {
    method: 'POST',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
    .catch(error => ({
      exception:        'Some error occurred',
      exceptionMessage: error
    }))
}

export async function updateUser({ id, datasetIds, email, name, role }) {
  const request = {
    'dataset_refs': datasetIds,
    emailAddress:   email,
    name,
    role
  }
  const url = '/api/users/' + id
  return await fetch(url, {
    method: 'PUT',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
    .catch(error => ({
      exception:        'Some error occurred',
      exceptionMessage: error
    }))
}

function groupBy(list, props) {
  return list.reduce((a, b) => {
    (a[b[props]] = a[b[props]] || []).push(b)
    return a
  }, {})
}
