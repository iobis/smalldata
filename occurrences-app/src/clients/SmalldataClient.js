import { format } from 'date-fns'

export async function getDatasets() {
  const response = await fetch('/api/datasets')
    .then(response => response.json())
  return response.map(renameRefToId)
}

export function renameRefToId({ ref, ...rest }) {
  return ({ id: ref, ...rest })
}

export function datasetTitleOf(dataset) {
  const title = dataset && dataset.title && dataset.title.value
  return title || ''
}

export function getOccurrenceMock() {
  return [{
    id:             1,
    addedDate:      '2011-12-01',
    scientificName: 'Abra Alba',
    dataset:        'NPPSD Short-tailed Albatross Sightings',
    occurrenceDate: '2011-12-09'
  }, {
    id:             2,
    addedDate:      '2011-12-01',
    scientificName: 'Abra Alba',
    dataset:        'NPPSD Short-tailed Albatross Sightings',
    occurrenceDate: '2011-12-09'
  }]
}

export async function postOccurrence({ occurence }) {
  const userRef = 'ovZTtaOJZ98xDDY'
  const occurrence = {
    core:       'occurrence',
    occurrence: [{
      tdwg: {
        datasetID:                     occurence.dataset.id,
        datasetName:                   occurence.dataset.title.value,
        basisOfRecord:                 occurence.occurrenceData.basisOfRecord.charAt(0).toUpperCase() + occurence.occurrenceData.basisOfRecord.slice(1),
        eventDate:                     format(occurence.occurrenceData.beginDate, 'YYYY-MM-DD') + '/' + format(occurence.occurrenceData.endDate, 'YYYY-MM-DD'),
        lifestage:                     occurence.occurrenceData.lifeStage,
        occurrenceStatus:              occurence.occurrenceData.occurrenceStatus,
        scientificName:                occurence.occurrenceData.scientificName,
        sex:                           occurence.occurrenceData.sex,
        decimalLongitude:              occurence.locationData.decimalLongitude,
        decimalLatitude:               occurence.locationData.decimalLatitude,
        coordinateUncertaintyInMeters: occurence.locationData.coordinateUncertainty,
        minimumDepthInMeters:          occurence.locationData.minimumDepth,
        maximumDepthInMeters:          occurence.locationData.maximumDepth,
        verbatimCoordinates:           occurence.locationData.verbatimCoordinates,
        verbatimDepth:                 occurence.locationData.verbatimDepth
      }
    }],
    emof:       [{
      purl:  {},
      iobis: {}
    }, {
      iobis: {}
    }]
  }
  const url = `/api/dwca/${occurence.dataset.id}/user/${userRef}/records`
  return await fetch(url, {
    method:  'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body:    JSON.stringify(occurrence)
  }).then(response => response.json())
}
