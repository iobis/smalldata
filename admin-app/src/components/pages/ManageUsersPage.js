import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import Divider from '../layout/Divider'
import { getUsersWithDatasets } from '@smalldata/dwca-lib/src/clients/SmalldataClient'

export default function ManageUsersPage() {
  const { t } = useTranslation()
  const [users, setUsers] = useState([])

  useEffect(() => {
    const fetchUsers = async() => {
      setUsers(await getUsersWithDatasets())
    }
    fetchUsers()
  }, [])

  return (
    <div className="manage-users-page">
      <section className="section">
        <div className="container is-fluid has-text-centered">
          <Link className="button is-info" to="/manage-users/create">
            {t('manageUsersPage.inputNew')}
          </Link>
        </div>
      </section>
      <Divider>{t('manageUsersPage.or')}</Divider>
      <section className="section">
        <div className="container is-fluid has-text-centered">
          <h4 className="title is-4">{t('manageUsersPage.managePreviousEntries')}</h4>
          <table className="table is-striped is-hoverable is-fullwidth">
            <thead>
              <tr>
                <th/>
                <th>{t('manageUsersPage.table.email')}</th>
                <th>{t('manageUsersPage.table.accessToDatasets')}</th>
                <th>{t('manageUsersPage.table.role')}</th>
              </tr>
            </thead>
            <tbody>
              {users.map(user => (
                <UserRow
                  datasets={user.datasets}
                  email={user.emailAddress}
                  id={user._id}
                  key={user._id}/>
              ))}
            </tbody>
          </table>
        </div>
      </section>
    </div>
  )
}

function UserRow({ id, email, datasets }) {
  const { t } = useTranslation()

  const toEdit = {
    pathname: '/manage-users/update/' + id,
    state:    {
      action:     'update',
      datasetIds: datasets.map(dataset => dataset.id),
      email,
      id,
      name:       '',
      role:       'researcher'
    }
  }

  return (
    <tr className="user-row">
      <td className="edit">
        <Link className="button is-info" to={toEdit}>
          {t('common.edit')}
        </Link>
      </td>
      <td className="email">{email}</td>
      <td className="datasets">
        <ul>
          {datasets.map(dataset => <li key={dataset.id}>{dataset.title.value}</li>)}
        </ul>
      </td>
      <td className="role"/>
    </tr>
  )
}

const datasetShape = {
  id:    PropTypes.string.isRequired,
  title: PropTypes.shape({
    value: PropTypes.string.isRequired
  }).isRequired
}

UserRow.propTypes = {
  datasets: PropTypes.arrayOf(PropTypes.shape(datasetShape)).isRequired,
  email:    PropTypes.string.isRequired,
  id:       PropTypes.string.isRequired
}
