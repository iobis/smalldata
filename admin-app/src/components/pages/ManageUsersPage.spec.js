import SmalldataClient from '@smalldata/dwca-lib/src/clients/SmalldataClient'
import ManageUsersPage from './ManageUsersPage'
import { MemoryRouter } from 'react-router-dom'
import { mount } from 'enzyme'
import React from 'react'
import { act } from 'react-dom/test-utils'

jest.mock('@smalldata/dwca-lib/src/clients/SmalldataClient', () => ({
  getUsersWithDatasets: jest.fn().mockImplementation(() =>
    new Promise((resolve) => {
      resolve([{
        _id:          'user-1',
        emailAddress: 'email-1@domain.com',
        datasets:     [{
          id:    'dataset-1-id',
          title: { value: 'dataset-1-title' }
        }]
      }])
    })
  )
}))

describe('ManageUsersPage', () => {
  const originalError = console.error
  let wrapper

  beforeAll(() => {
    console.error = (...args) => {
      if (/Warning.*not wrapped in act/.test(args[0])) return
      originalError.call(console, ...args)
    }
  })

  it('renders correctly for empty users', async() => {
    act(() => {
      wrapper = mount(
        <MemoryRouter initialEntries={[{ pathname: '/manage-users', key: 'testKey' }]}>
          <ManageUsersPage/>
        </MemoryRouter>
      )
    })
    expect(SmalldataClient.getUsersWithDatasets).toHaveBeenCalledWith()

    await flushPromises()
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('tbody tr')).toHaveLength(1)
  })
})

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve))
}
