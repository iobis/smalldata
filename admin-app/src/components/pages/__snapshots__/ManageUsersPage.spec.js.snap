// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageUsersPage when users provided renders correctly 1`] = `
<ManageUsersPage>
  <div
    className="manage-users-page"
  >
    <section
      className="section"
    >
      <div
        className="container is-fluid has-text-centered"
      >
        <Link
          className="button is-info"
          to="/manage-users/create"
        >
          <a
            className="button is-info"
            href="/manage-users/create"
            onClick={[Function]}
          >
            manageUsersPage.inputNew
          </a>
        </Link>
      </div>
    </section>
    <Divider>
      <div
        className="divider"
      >
        <span>
          manageUsersPage.or
        </span>
      </div>
    </Divider>
    <section
      className="section"
    >
      <div
        className="container is-fluid has-text-centered"
      >
        <h4
          className="title is-4"
        >
          manageUsersPage.managePreviousEntries
        </h4>
        <table
          className="table is-striped is-hoverable is-fullwidth"
        >
          <thead>
            <tr>
              <th />
              <th>
                manageUsersPage.table.email
              </th>
              <th>
                manageUsersPage.table.accessToDatasets
              </th>
              <th>
                manageUsersPage.table.role
              </th>
            </tr>
          </thead>
          <tbody>
            <UserRow
              datasets={
                Array [
                  Object {
                    "id": "dataset-1-id",
                    "title": Object {
                      "value": "dataset-1-title",
                    },
                  },
                ]
              }
              email="email-1@domain.com"
              id="user-1"
              key="user-1"
            >
              <tr
                className="user-row"
              >
                <td
                  className="edit"
                >
                  <Link
                    className="button is-info"
                    to={
                      Object {
                        "pathname": "/manage-users/update/user-1",
                        "state": Object {
                          "action": "update",
                          "datasetIds": Array [
                            "dataset-1-id",
                          ],
                          "email": "email-1@domain.com",
                          "id": "user-1",
                          "name": "",
                          "role": "researcher",
                        },
                      }
                    }
                  >
                    <a
                      className="button is-info"
                      href="/manage-users/update/user-1"
                      onClick={[Function]}
                    >
                      common.edit
                    </a>
                  </Link>
                </td>
                <td
                  className="email"
                >
                  email-1@domain.com
                </td>
                <td
                  className="datasets"
                >
                  <ul>
                    <li
                      key="dataset-1-id"
                    >
                      dataset-1-title
                    </li>
                  </ul>
                </td>
                <td
                  className="role"
                >
                  â€”
                </td>
              </tr>
            </UserRow>
            <UserRow
              datasets={
                Array [
                  Object {
                    "id": "dataset-1-id",
                    "title": Object {
                      "value": "dataset-1-title",
                    },
                  },
                ]
              }
              email="email-2@domain.com"
              id="user-2"
              key="user-2"
              role="node manager"
            >
              <tr
                className="user-row"
              >
                <td
                  className="edit"
                >
                  <Link
                    className="button is-info"
                    to={
                      Object {
                        "pathname": "/manage-users/update/user-2",
                        "state": Object {
                          "action": "update",
                          "datasetIds": Array [
                            "dataset-1-id",
                          ],
                          "email": "email-2@domain.com",
                          "id": "user-2",
                          "name": "",
                          "role": "node manager",
                        },
                      }
                    }
                  >
                    <a
                      className="button is-info"
                      href="/manage-users/update/user-2"
                      onClick={[Function]}
                    >
                      common.edit
                    </a>
                  </Link>
                </td>
                <td
                  className="email"
                >
                  email-2@domain.com
                </td>
                <td
                  className="datasets"
                >
                  <ul>
                    <li
                      key="dataset-1-id"
                    >
                      dataset-1-title
                    </li>
                  </ul>
                </td>
                <td
                  className="role"
                >
                  node manager
                </td>
              </tr>
            </UserRow>
          </tbody>
        </table>
      </div>
    </section>
  </div>
</ManageUsersPage>
`;
